crates/pathfinder/src/state/class_hash.rs

```diff
@@ -117,39 +117,75 @@ fn compute_class_hash0(mut contract_definition: json::ContractDefinition<'_>) ->
+    fn add_extra_space_to_cairo_named_tuples(value: &mut serde_json::Value) {
+        match value {
+            serde_json::Value::Array(v) => walk_array(v),
+            serde_json::Value::Object(m) => walk_map(m),
+            _ => {}
+        }
+    }
+
+    fn walk_array(array: &mut Vec<serde_json::Value>) {
+        for v in array.iter_mut() {
+            add_extra_space_to_cairo_named_tuples(v);
+        }
+    }
+
+    fn walk_map(object: &mut serde_json::Map<String, serde_json::Value>) {
         for (k, v) in object.iter_mut() {
-            if k == "cairo_type" {
-                // Check that the value is a tuple.
-                if let Some(tuple_inner) = v
-                    .as_str()
-                    .and_then(|v| v.strip_prefix('(').and_then(|v| v.strip_suffix(')')))
+            match v {
+                serde_json::Value::String(s) => {
+                    let new_value = add_extra_space_to_named_tuple_type_definition(&k, &s);
+                    if new_value.as_ref() != s {
+                        *v = serde_json::Value::String(new_value.into());
+                    }
+                }
+                _ => add_extra_space_to_cairo_named_tuples(v),
+            }
+        }
+    }
+
+    fn add_extra_space_to_named_tuple_type_definition<'a>(
+        key: &str,
+        value: &'a str,
+    ) -> std::borrow::Cow<'a, str> {
+        use std::borrow::Cow::*;
+        match key {
+            "cairo_type" => {
+                if let Some(tuple_inner) = value.strip_prefix('(').and_then(|v| v.strip_suffix(')'))
                 {
-                    let new_val = format!(
-                        "({})",
-                        // FIXME: Replace this crude hack with something more intelligent.
-                        //        It is required because if we receive an already correct ` : `, we will still
-                        //        "repair" it to `  : ` which we then fix at the end.
-                        tuple_inner.replace(": ", " : ").replace("  :", " :")
-                    );
-                    *v = serde_json::Value::String(new_val);
+                    let new_val = format!("({})", add_extra_space_before_colon(tuple_inner));
+                    Owned(new_val)
+                } else {
+                    Borrowed(value)
                 }
-            } else if let Some(inner_object) = v.as_object_mut() {
-                // `cairo_type` occurs at multiple levels in the object, so we must explore deeper.
-                insert_space(inner_object);
             }
+            "value" => {
+                if let Some(cast_inner) = value
+                    .strip_prefix("[cast(")
+                    .and_then(|v| v.strip_suffix(")]"))
+                {
+                    let new_val = format!("[cast({})]", add_extra_space_before_colon(cast_inner));
+                    Owned(new_val)
+                } else {
+                    Borrowed(value)
+                }
```

Comment:
I wonder if this tuple can only occur within the `[cast(..)]` type framework. Should we not just be applying the formatting regardless? Similarly with the `cairo_type`.